src/control/BBTrajectories/BBTrajectory1D.tpp

```diff
@@ -0,0 +1,210 @@
+    numParts = 2;
+}
+template<class num>
+void BBTrajectory1D<num>::updatePart(int index, num tEnd, num acc, num vel, num pos) noexcept {
+    parts[index].acc = acc;
+    parts[index].startPos = pos;
+    parts[index].startVel = vel;
+    parts[index].tEnd = tEnd;
+}
+template<class num>
+void BBTrajectory1D<num>::trapezoidalProfile(num initialPos, num initialVel, num maxVel, num finalPos,
+        num maxAcc) noexcept {
+    num acc1;
+    num acc3;
+
+    if (initialVel > maxVel) {
+        acc1 = - maxAcc;
+    }
+    else {
+        acc1 = maxAcc;
+    }
+
+    if (maxVel > 0) {
+        acc3 = - maxAcc;
+    }
+    else {
+        acc3 = maxAcc;
+    }
+
+    num t1 = (maxVel - initialVel)/acc1;
+    num t3 = - maxVel/acc3;
+    num startCoastPos = initialPos + 0.5*(initialVel + maxVel)*t1;
+    num endCoastPos = finalPos - 0.5*maxVel*t3;
+    num t2 = (endCoastPos - startCoastPos)/maxVel;
+
+    updatePart(0, t1, acc1, initialVel, initialPos);
+    updatePart(1, t1 + t2, 0, maxVel, startCoastPos);
+    updatePart(2, t1 + t2 + t3, acc3, maxVel, endCoastPos);
+    numParts = 3;
+
+}
+template<class num>
+void BBTrajectory1D<num>::generateTrajectory(num initialPos, num initialVel, num finalPos, num maxVel,
+        num maxAcc) noexcept {
+    num brakePos = fullBrakePos(initialPos, initialVel, maxAcc);
+    if (brakePos <= finalPos) {
+        //Check if we need triangular profile or trapezoidal:
+        num accBrakePos = accelerateBrakePos(initialPos, initialVel, maxVel, maxAcc);
+        if (accBrakePos >= finalPos) {
+            triangularProfile(initialPos, initialVel, finalPos, maxAcc, true);
+        }
+        else {
+            trapezoidalProfile(initialPos, initialVel, maxVel, finalPos, maxAcc);
+        }
+    }
+    else {
+        //similar to above but with slightly mirrored logic
+        num accBrakePos = accelerateBrakePos(initialPos, initialVel, - maxVel, maxAcc);
+        if (accBrakePos <= finalPos) {
+            triangularProfile(initialPos, initialVel, finalPos, maxAcc, false);
+        }
+        else {
+            trapezoidalProfile(initialPos, initialVel, - maxVel, finalPos, maxAcc);
+        }
+    }
+}
+template<class num>
+BBTrajectory1D<num>::BBTrajectory1D(num initialPos, num initialVel, num finalPos, num maxVel, num maxAcc) noexcept
+        :
+        m_initialPos{initialPos},
+        m_initialVel{initialVel},
+        m_finalPos{finalPos},
+        m_maxAcc{maxAcc},
+        m_maxVel{maxVel} {
+    generateTrajectory(initialPos, initialVel, finalPos, maxVel, maxAcc);
+}
+template<class num>
+PosVelAcc<num> BBTrajectory1D<num>::getValues(num t) const noexcept {
+    num trajTime = fmax(0, t);
+    part piece = parts[0];
+    if (trajTime >= getTotalTime()) {
+        //The time is not on the trajectory so we just return the last known element
+        return PosVelAcc<num>(m_finalPos, 0, 0);//can also be computed from parts if necessary
```

Comment:
```suggestion
        return PosVelAcc{ m_finalPos, 0, 0 }; //can also be computed from parts if necessary
```

Reply:
Can't. I made the constructor explicit to prevent clang from whining.